File: easyfamily-ios/EasyfamilyApp/Features/Finance/LiabilityViewModel.swift
import Foundation

@MainActor
final class LiabilityViewModel: ObservableObject {
    @Published var liabilities: [LiabilityItem] = []
    @Published var totalBalance: Decimal = 0
    @Published var totalMonthlyPayment: Decimal = 0
    @Published var isLoading: Bool = false
    @Published var error: String?

    func load(token: String) async {
        isLoading = true
        error = nil
        do {
            let response = try await APIService.listLiabilities(token: token)
            liabilities = response.items
            totalBalance = response.totalBalance
            totalMonthlyPayment = response.totalMonthlyPayment
        } catch {
            self.error = "加载失败：\(error.localizedDescription)"
        }
        isLoading = false
    }

    func create(token: String, name: String, liabilityType: String, balance: Decimal,
                monthlyPayment: Decimal?, interestRate: Decimal?, note: String?) async throws {
        let req = LiabilityCreateRequest(
            name: name, liabilityType: liabilityType, balance: balance,
            monthlyPayment: monthlyPayment, interestRate: interestRate,
            note: note?.isEmpty == false ? note : nil
        )
        _ = try await APIService.createLiability(token: token, req: req)
        await load(token: token)
    }

    func update(token: String, id: Int, name: String, liabilityType: String, balance: Decimal,
                monthlyPayment: Decimal?, interestRate: Decimal?, note: String?) async throws {
        let req = LiabilityCreateRequest(
            name: name, liabilityType: liabilityType, balance: balance,
            monthlyPayment: monthlyPayment, interestRate: interestRate,
            note: note?.isEmpty == false ? note : nil
        )
        _ = try await APIService.updateLiability(token: token, id: id, req: req)
        await load(token: token)
    }

    func delete(token: String, id: Int) async throws {
        try await APIService.deleteLiability(token: token, id: id)
        await load(token: token)
    }
}

File: easyfamily-android/app/src/main/java/com/easyfamily/ui/liability/LiabilityViewModel.kt
package com.easyfamily.ui.liability

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.easyfamily.data.ApiClient
import com.easyfamily.data.LiabilityData
import com.easyfamily.data.local.AuthDataStore
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import javax.inject.Inject

data class LiabilityUiState(
    val liabilities: List<LiabilityData> = emptyList(),
    val totalBalance: Double = 0.0,
    val totalMonthlyPayment: Double = 0.0,
    val loading: Boolean = false,
    val error: String? = null
)

@HiltViewModel
class LiabilityViewModel @Inject constructor(
    private val authDataStore: AuthDataStore
) : ViewModel() {

    private val _uiState = MutableStateFlow(LiabilityUiState())
    val uiState: StateFlow<LiabilityUiState> = _uiState.asStateFlow()

    private suspend fun token(): String = authDataStore.accessToken.first()

    fun load() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(loading = true, error = null)
            try {
                val t = token()
                val (items, total, monthly) = ApiClient.listLiabilities(t)
                _uiState.value = _uiState.value.copy(
                    liabilities = items,
                    totalBalance = total,
                    totalMonthlyPayment = monthly,
                    loading = false
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(loading = false, error = e.message)
            }
        }
    }

    fun createLiability(
        name: String,
        liabilityType: String,
        balance: Double,
        monthlyPayment: Double?,
        interestRate: Double?,
        note: String?
    ) {
        viewModelScope.launch {
            try {
                ApiClient.createLiability(token(), name, liabilityType, balance, monthlyPayment, interestRate, note)
                load()
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.message)
            }
        }
    }

    fun updateLiability(
        id: Int,
        name: String,
        liabilityType: String,
        balance: Double,
        monthlyPayment: Double?,
        interestRate: Double?,
        note: String?
    ) {
        viewModelScope.launch {
            try {
                ApiClient.updateLiability(token(), id, name, liabilityType, balance, monthlyPayment, interestRate, note)
                load()
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.message)
            }
        }
    }

    fun deleteLiability(id: Int) {
        viewModelScope.launch {
            try {
                ApiClient.deleteLiability(token(), id)
                load()
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.message)
            }
        }
    }
}
